### PR TITLE
hoist list

### PR DESCRIPTION
**Pull request by blaggacao**
- refactor: let tranformers take a cursor to reason about depth
- feat: optionally accept transformer list
- docs: add configurable hoistLists transformer
- test: add configurable hoistLists transformer
- feat: add configurable hoistLists transformer

### CODE REVIEW

File: src/load.nix [hunk 1]
```diff
@@ -121,24 +122,34 @@ in
 { src
 , loader ? root.loaders.default
 , inputs ? { }
-, transformer ? id
+, transformer ? _: id
```
Comment (figsoda) [inline src/load.nix:125]: please also update the documentation for `load` and `transformers.liftDefault`

File: src/load.nix [hunk 2]
```diff
@@ -121,24 +122,34 @@ in
...
 }:
+let
+  transformer' =
+    if isList transformer
+    then
+      (cursor: mod: pipe mod (
+        map (t: t cursor) transformer)
+      )
+    else transformer;
```
Comment (figsoda) [inline src/load.nix:134]:
```suggestion
  transformer' = cursor: flip pipe
    (map (t: t cursor) (toList transformer));
```

File: src/transformers/hoistLists.nix [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  /*mergeAttrsButConcatOn: merge attributes shallowly, but
+    concat values of a specific key into a list in that key
+
+    Type:
+      (key :: String -> AttrSet -> AttrSet) -> { ${key} :: List; ... }
+
+    */
+  mergeAttrsButConcatOn = key: x: y:
+    x // y // {
+      ${key} = flatten (catAttrs key [ x y ]);
```
Comment (figsoda) [inline src/transformers/hoistLists.nix:32]: `concatLists` should be enough here

```suggestion
      ${key} = concatLists (catAttrs key [ x y ]);
```

File: src/transformers/hoistLists.nix [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      concatMapAttrsWith (mergeAttrsButConcatOn "mykey")
+        (name: value: {
+          ${name} = value;
+          ${key} = value ++ value;
+        })
+        { x = "a"; y = "b"; }
+      => { x = "a"; y = "b"; mykey = [ "aa" "bb"]; }
+
+    */
+  concatMapAttrsWith = mergeFun: f: flip pipe [ (mapAttrs f) attrValues (foldl' mergeFun { }) ];
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/transformers/hoistLists.nix:53]: A `_utils.nix` file :+1: :smile:

File: src/transformers/hoistLists.nix [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+in
+cursor: mod:
+if cursor == [ ] # toplevel
+then
+  concatMapAttrsWith (mergeAttrsButConcatOn to)
+    (file: value: if ! value ? ${from} then { ${file} = value; } else {
+      ${file} = removeAttrs value [ from ];
+      ${to} = value.${from};
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/transformers/hoistLists.nix:62]: Yes, it's intentional. I'll clarify!

**Comment by figsoda** [comment]
btw some suggestions I made requires updating the `let inherit`s

**Comment by blaggacao** [comment]
Ok, I think I got all. Can you review again?

**Pull request merged**